### PR TITLE
Add linting and unit testing job to CI for Dagster code

### PR DESCRIPTION
**Pull request by ThePletch**
New features:
* Adds a Github action to automatically evaluate our Python code for Dagster
  * Runs our test suite automatically via Pytest
  * Lint-checks our code according to the PEP8 style guide
  * Evaluates our code's type hints to ensure they all match up (and all of our first-party code has hints).
* Adds a pre-commit hook to automatically lint Python code as it's committed.

### CODE REVIEW

File: orchestration/dagster_orchestration/hca_orchestration/contrib/argo_workflows.py [hunk 1]
```diff
@@ -16,9 +17,20 @@ def generate_argo_archived_workflows_client(host_url: str, access_token: str) ->
             header_value=f"Bearer {access_token}"))
 
 
-class ArgoArchivedWorkflowsClientMixin:
-    def __init__(self, *args, argo_url: str, access_token: str, **kwargs):
-        super().__init__(*args, **kwargs)
+# protocols let us set up complex type constraints on function arguments.
+# this is a useful alternative when referencing functions from third-party libraries that aren't
+# properly type annotated.
+class ArgoFetchListOperation(Protocol):
```
Comment (aherbst-broad) [inline orchestration/dagster_orchestration/hca_orchestration/contrib/argo_workflows.py:23]: 👍 Handy, I did know about that PEP.

File: orchestration/dagster_orchestration/hca_orchestration/sensors.py [hunk 2]
```diff
@@ -4,15 +4,15 @@
...
 from dagster import sensor, RunRequest, SkipReason
 
-from hca_orchestration.contrib.argo_workflows import ArgoArchivedWorkflowsClientMixin, ExtendedArgoWorkflow
+from hca_orchestration.contrib.argo_workflows import ArgoArchivedWorkflowsClient, ExtendedArgoWorkflow
 from hca_orchestration.resources.base import default_google_access_token
 
 
 # boundary before which we don't care about any workflows in argo
-ARGO_EPOCH = datetime(2021, 3, 1)
+ARGO_EPOCH: datetime = datetime(2021, 3, 1)
```
Comment (aherbst-broad) [inline orchestration/dagster_orchestration/hca_orchestration/sensors.py:12]: I still think it's worth bumping this.

File: orchestration/dagster_orchestration/hca_manage/manage.py [hunk 3]
```diff
@@ -259,9 +293,9 @@ def delete_dataset(self, dataset_name: Optional[str] = None, dataset_id: Optiona
         else:
             # can't have both/neither provided
             raise ValueError("You must provide either dataset_name or dataset_id, and cannot provide neither/both.")
-        response = self.data_repo_client.delete_dataset(dataset_id)
-        logging.info(f"Dataset deletion job id: {response.id}")
-        return response.id
+        delete_response = self.data_repo_client.delete_dataset(dataset_id)
```
Comment (aherbst-broad) [inline orchestration/dagster_orchestration/hca_manage/manage.py:296]: ⛺ Nice rename

File: orchestration/dagster_orchestration/hca_manage/manage.py [hunk 4]
```diff
@@ -40,18 +40,23 @@ def __init__(self, environment: str, data_repo_client: RepositoryApi, project: O
...
             "dev": ["hca-snapshot-readers@dev.test.firecloud.org"],
             "prod": ["hca-snapshot-readers@firecloud.org"]
         }[environment]
 
+        self._gcp_creds = None
+
+    def gcp_creds(self):
+        # use application default credentials to seamlessly work across monster devs
+        # assumes `gcloud auth application-default login` has been run
+        if not self._gcp_creds:
```
Comment (aherbst-broad) [inline orchestration/dagster_orchestration/hca_manage/manage.py:53]: ❓ Why shift to lazy-loading the app creds?

File: orchestration/dagster_orchestration/hca_orchestration/sensors.py (hunk 2 above)
Comment (ThePletch) [reply orchestration/dagster_orchestration/hca_orchestration/sensors.py:12]: agreed, going to do it in the other pr before merging

File: orchestration/dagster_orchestration/hca_manage/manage.py (hunk 4 above)
Comment (ThePletch) [reply orchestration/dagster_orchestration/hca_manage/manage.py:53]: eager loading them causes authentication errors in the tests on initialization, even if we've mocked out all external api calls, whereas lazy loading doesn't try to call out to gcloud unless we are trying to make a live gcloud request

**Review by aherbst-broad — COMMENTED** [review]
Looking good, it appears that checks are failing though.

File: orchestration/dagster_orchestration/README.md [hunk 5]
```diff
@@ -4,9 +4,15 @@ This is an initial introduction of [Dagster](https://dagster.io)
 into our codebase for workflow orchestration. This document captures notes,
 todos and anything else we bump into as we test out this technology.
 
+## Setting up your dev environment
+* Make sure you have [Poetry](https://python-poetry.org/docs/#installation) installed.
+* From the `dagster_orchestration` dir, run `poetry install` to install all dependencies.
+* Run `poetry run pre-commit install` to install a hook that will automatically lint your code whenever you create a commit.
```
Comment (aherbst-broad) [inline orchestration/dagster_orchestration/README.md:10]: 👍 another TIL, thanks for sharing.

**Review by aherbst-broad — APPROVED** [review]

**Review by raaidbroad — APPROVED** [review]

**Pull request merged**